theano/gof/tests/test_wrapper.py

```diff
@@ -0,0 +1,245 @@
+from __future__ import absolute_import, print_function, division
+import theano
+import numpy
+from unittest import TestCase
+from theano.gof import Op, COp, Apply
+from theano import Generic
+from theano.scalar import Scalar
+from theano.tensor import TensorType
+from theano.gof.wrapper import Wrapper, Wrap
+from theano import config
+from theano import tensor
+from theano.tests import unittest_tools as utt
+
+dtype = config.floatX
+tensor_type_0d = TensorType(dtype, tuple())
+scalar_type = Scalar(dtype)
+generic_type = Generic()
+
+
+# A test op to compute `y = a*x^2 + bx + c` for any tensor x, with a, b, c as op params.
+class QuadraticOpFunc(Op):
+    __props__ = ('a', 'b', 'c')
+    params_type = Wrapper(a=tensor_type_0d,
+                          b=scalar_type,
+                          c=generic_type)
```

Comment:
There is something I don't like in this logic. First the dtype is fixed for the class instance. We will have cases where the same op instance will be used in many apply nodes that have different inputs dtypes. We need to cover that cases.

I see those options:

- Have the params_type be created in \__init__. I don't like it much, we it will force the dtype to be know before the init.
- Have the params_type allow to be a callable/class method that will be called with the node and when called, should return a Wrapper instance.

Where do we need to have this wrapper? If we need to create it at only one place, a class method seem good. But if we need to call the method at many places, we will need to include that in the Apply node I think (or accept that it get called many times, not my preference)